src/parser.ts

```diff
@@ -0,0 +1,264 @@
+    const result = this.expect(this.grammar.begin());
+    result.errorPos = this.errorPos;
+    return result;
+  }
+
+  matchFail(returnPos: number): void {
+    if (this.errorPos === NO_ERROR_POS) this.errorPos = this.pos;
+    this.pos = returnPos;
+  }
+
+  matchSuccess(): void {
+    this.errorPos = NO_ERROR_POS;
+  }
+
+  // Look for `token` in this.line[this.pos], and return success
+  // if one is found. `token` is specified as a regular expression.
+  matchToken(token: string, noSpace: boolean = false): Result {
+    const regexp = noSpace ? new RegExp('^((' + token + '))') : new RegExp('^((' + token + ')\\s*)');
+    const workingLine = this.line.slice(this.pos);
+    const result = workingLine.match(regexp);
+    if (result !== null) {
+      return {
+        success: true,
+        matchedString: result[2],
+        incrementPos: result[1].length,
+        pos: this.pos,
+      };
+    } else {
+      return {
+        success: false,
+        pos: this.pos,
+      };
+    }
+  }
+
+  // Execute rule to match a sequence of tokens (or rules). If `maybe` is
+  // set, then return success even if the token is not found, but reset
+  // the position before exiting.
+  // TODO: expectOne(...) is never called with the `maybe` parameter.
+  expectOne(rule: Rule, maybe: boolean = false): Result {
+    const results: Result[] = [];
+    const pos = this.pos;
+
+    let allMatches = true;
+    let oneMatch = false;
+    maybe = maybe === true || rule.maybe === true;
+
+    // Execute all sub rules in sequence.
+    if (rule.expect) {
+      for (const next of rule.expect) {
+        const localPos = this.pos;
+        const result = this.expect(next);
+
+        // If `rule.or` is set, then return success if any one
+        // of the subrules match, else all subrules must match.
+        if (result.success) {
+          results.push(result);
+          oneMatch = true;
+          if (rule.or) break;
+        } else {
+          allMatches = false;
+          if (!rule.or) {
+            this.pos = localPos;
+            break;
+          }
+        }
+      }
+    }
+
+    const gotOne = (rule.or && oneMatch) || allMatches;
+    const success = gotOne || maybe === true;
+    const numMatches = gotOne ? 1 : 0;
+    if (maybe && !gotOne) this.pos = pos;
+    if (success) {
+      this.matchSuccess();
+    } else {
+      this.matchFail(pos);
+    }
+    return { success, results, numMatches };
+  }
+
+  // Try to match multiple (one or more) instances of the rule. If `maybe` is set,
+  // then a failed match is also a success (but the position is reset).
+  expectOneOrMore(rule: Rule, maybe: boolean = false): Result {
+    const results: Result[] = [];
+    const pos = this.pos;
+    let numMatches = 0;
+    let more = true;
+
+    do {
+      const result = this.expectOne(rule);
+      if (result.success && result.results) {
+        numMatches++;
+        // TODO: Is it okay to use the spread operator here to flatten the results?
```

Comment:
Yes, this is fine I think. I'd suggest adding a few more tests to `tests/parser_tests.js` to exercise this path, and make sure the before/after behavior is correct.

Reply:
Sounds good. Will write some tests.

Reply:
Hi @ronyeh -- did you manage to figure out tests? Can I help in any way?

Reply:
Is there a way to just run a single test suite, via npm / grunt options? I'm working on a parser test now and it seems easier for me to write it separately on my own demo site, and then copy it into `parser_tests.js` when I'm done.

To run a single QUnit test, I've been doing this in `run.js`:

```
VF.Test.run = function () {
  VF.Test.Parser.Start(); // Only run this test!
  // skips everything else, but doesn't trigger any eslint / TS compiler errors.
  if (Math.random() > -1) {
    return;
  }
 ... 
```

Reply:
Not sure about the command line, but my workflow includes keeping the test page open in the browser and using the dropdown to select a specific test category (e.g., parser). I use `grunt watch` to rebuild as I change code, and just refresh the page. It's quick and effective.

Reply:
Thanks for the workflow tips! I didn't even think to open flow.html locally in my browser, and I also did not notice that dropdown in the upper right.

I have been using my own test pages running on a localhost webserver, and I also watch the vexflow library with:
`grunt webpack:watch`

`grunt watch` seems to update the tests, but if I change the vexflow core library, it doesn't update the build files (even though it triggers the grunt watch and concat:tests task since src/* is monitored).

Is grunt watch supposed to also recompile the vexflow library + vexflow tests?

Reply:
Also, flow.html has an outdated title:
**VexFlow 2** - JavaScript Music Notation and Guitar Tab

We could change it to:
VexFlow - Music Notation and Guitar Tab

...and it'll always be up to date, even when you upgrade to VexFlow 4!

I won't submit a PR for this, since you can just change it yourself on master, haha.

Reply:
@rvilarl  have you been using `grunt watch` successfully? I noticed that it concats & outputs `vexflow-tests.js` and 
`vexflow-tests.js.map` into the vexflow root directory, which seems incorrect.

So I modified Gruntfile.js to look like this:
```
const TARGET_TESTS_BROWSER = 'build/vexflow-tests.js'; // I changed it to output to the build folder
```
previously it looked like this:
```
const TARGET_TESTS_BROWSER = 'vexflow-tests.js'; // outputs to the vexflow root dir
```

So now when I load flow.html, it loads the correct vexflow-tests.js file.

However, I get this error:
```
vexflow-tests.js:16989 Uncaught SyntaxError: Cannot use import statement outside a module
flow.html:26 Uncaught ReferenceError: VF is not defined
    at HTMLDocument.<anonymous> (flow.html:26)

```

I think it is because we are using grunt-contrib-concat to combine the test files. However, some of the test files (e.g. music_tests.ts) now have import statements because they are in TypeScript.

I think we need to modify the watch tasks to use webpack to compile the TS..... alas I'm no expert in grunt or webpack.

Reply:
I am now able to dev and test quickly by refreshing flow.html in the browser.

I added a watch task that will watch over tests/, and trigger the webpack build.

To make it faster, I run the webpack task for src/ and the webpack task for tests/ in parallel. So if you only change a file in tests/, it won't rebuild the vexflow core library.

See:
https://github.com/0xfe/vexflow/pull/961

Reply:
@ronyeh I have never used `grunt watch` :) Now I just use:
- `npm run reference` on master to produce the reference images
- `npm run test:reference` on the branches to make sure that there are no diff on images compared to master
- and `flow.html` with Chrome to debug.

Reply:
It's possible that `grunt watch` broke during the typescript infra change. Thanks for the PR, looking now.